src/methods/widen.js

```diff
@@ -210,15 +210,44 @@ export class WidenImplementation {
         }
         invariant(d2 !== undefined);
       }
-      return this.widenDescriptors(realm, d1, d2);
+      let result = this.widenDescriptors(realm, d1, d2);
+      if (result && result.value instanceof AbstractValue && result.value.kind === "widened") {
+        let rval = result.value;
+        let pathNode = b.pathNode;
+        if (pathNode === undefined) {
+          //Since properties already have mutable storage locations associated with them, we do not
+          //need phi nodes. What we need is an abstract value with a build node that results in a memberExpression
+          //that resolves to the storage location of the property.
+
+          // For now, we only handle loop invariant properties
+          //i.e. properties where the member expresssion does not involve any values written to inside the loop.
+          //todo: handle the case where key is an abstract value.
+          let key = b.key;
+          if (typeof key === "string") {
+            rval._buildNode = ([o]) => t.memberExpression(o, t.identifier(key));
```

Comment:
You mutate `rval` here. I don't like that. Why is it guaranteed that `rval` is a fresh temporal abstract value, and is there some implicit invariant that widened abstract values shall never be involved in some kind of CSE?

Reply:
rval is freshly constructed via AbstractValue.createFromWidening. The modification here can be regarded as completing the construction. This is very convenient, but yes, you are right that it is a bit of a code smell. I'll add another constructor method to AbstractValue.